Update GeoPoint FieldData for GeoPointV2

nknize opened the pull request:
This is GeoPointV2 PR part 1. It adds an abstraction layer to GeoPoint FieldData for supporting the new GeoPointField type in Lucene 5.4 along with backwards compatibility for legacy geo_point field types. This abstraction is necessary for 2.x only and will not be carried forward in 3.0.

s1monw commented:
@mikemccand @colings86 can you both please take a look at this?

colings86 commented on core/src/main/java/org/elasticsearch/index/fielddata/plain/AbstractGeoPointDVIndexFieldData.java:
```diff
@@ -46,29 +50,41 @@ public final XFieldComparatorSource comparatorSource(@Nullable Object missingVal
+                return (indexCreatedBefore2x) ? new GeoPointLegacyDVAtomicFieldData(DocValues.getBinary(context.reader(), fieldNames.indexName())) :
+                        new GeoPointDVAtomicFieldData(DocValues.getSortedNumeric(context.reader(), fieldNames.indexName()));
+            } catch (IOException e) {
+                throw new IllegalStateException("Cannot load doc values", e);
```
Is there anyway of outputting the name(s) of the fields we couldn't load doc values for here? Probably useful for debugging

colings86 commented:
@nknize Not sure how feasible it is but it would be good to have a test which ensures that geopoint doc values work correctly under the following scenarios:
* pre-2.2 index with only pre 2.2 segments
* pre 2.2 index with a mix of pre and post 2.2. segments
* post 2.2 index

Maybe we could make a backwards compatibility test work to do this?

mikemccand commented on core/src/main/java/org/elasticsearch/index/fielddata/plain/GeoPointArrayIndexFieldData.java:
```diff
@@ -72,12 +82,66 @@ public AtomicGeoPointFieldData loadDirect(LeafReaderContext context) throws Exce
+            Long hashedPoint;
+            long numTerms = 0;
+            while ((hashedPoint = iter.next()) != null) {
+                indexedPoints = BigArrays.NON_RECYCLING_INSTANCE.resize(indexedPoints, numTerms + 1);
```
Shouldn't this be `.oversize` not `.resize`?

mikemccand replied:
Oh I see: I think `.resize` is OK here since `BigLongArray` is paged ...

mikemccand commented on core/src/main/java/org/elasticsearch/index/fielddata/plain/AbstractGeoPointDVIndexFieldData.java:
```diff
@@ -46,29 +50,41 @@ public final XFieldComparatorSource comparatorSource(@Nullable Object missingVal
+        @Override
+        public AtomicGeoPointFieldData load(LeafReaderContext context) {
+            try {
+                return (indexCreatedBefore2x) ? new GeoPointLegacyDVAtomicFieldData(DocValues.getBinary(context.reader(), fieldNames.indexName())) :
```
Can you use a traditional if statement (not ternary operator) here?

mikemccand commented on core/src/main/java/org/elasticsearch/index/fielddata/plain/GeoPointArrayAtomicFieldData.java:
```diff
@@ -0,0 +1,147 @@
+                        final int ord = singleOrds.getOrd(docID);
+                        if (ord >= 0) {
+                            point.resetFromIndexHash(indexedPoints.get(ord));
+                        }
```
This is a little spooky: should we throw an exception in the else clause here?  Else the get of a `docId` that didn't index a geo point will silently return whatever `point` previously had?

nknize commented:
Addressed review comments. Once these current 3 PRs are merged I will have a final PR that cuts over to the GeoPointFieldV2. That PR will finish adding backcompat testing to existing unit and integration tests. I've also opened a new issue #14562 to address @colings86 point for having better backcompat testing for FieldData.

nknize commented:
@colings86 @mikemccand this should be ready. Let me know if you have any objections to the updates. If there are no objections I'd like to cut over to GeoPointV2 at some point today.

colings86 commented:
This LGTM although I'll leave the final LGTM to @mikemccand :)

mikemccand commented:
LGTM2!
